pkg/provider/ibmcloud/plugin/loadbalancer/lb.go

```diff
@@ -0,0 +1,466 @@
+	if err != nil {
+		return nil, err
+	}
+
+	// Iterate over the list of members to add checking if they already exist in the loadbalancer.
+	// If not, add member.
+	for _, id := range ids {
+		ip := string(id)
+		ipFound := false
+		for _, member := range lb.Members {
+			if *member.Address == ip {
+				ipFound = true
+				lbLogger.Error("RegisterBackends", "name", l.name, "V", debugV, "backend already exists", ip)
+				break
+			}
+		}
+		if ipFound == false {
+			member := datatypes.Network_LBaaS_LoadBalancerServerInstanceInfo{
+				PrivateIpAddress: &ip,
+				Weight:           &weight,
+			}
+			membersToAdd = append(membersToAdd, member)
+		}
+	}
+
+	// If there are any members to add, add them.
+	if len(membersToAdd) > 0 {
+		attempt := 0
+		for {
+			attempt++
+			_, err = l.lbMemberService.AddLoadBalancerMembers(&l.uuid, membersToAdd)
+			if err == nil {
+				break
+			}
+			// Check for the lb is in state UPDATE_PENDING. (HTTP 500)
+			if strings.Contains(err.Error(), "UPDATE_PENDING") && attempt <= updateRetryCount {
+				time.Sleep(updateRetryTime)
+			} else {
+				return nil, err
+			}
+		}
+	}
+
+	lbLogger.Debug("RegisterBackends", "name", l.name, "V", debugV, "backends", membersToAdd)
+	return lbResult("register"), nil
+}
+
+// DeregisterBackend removes the specified instances from the backend pool
+func (l *ibmcloudlb) DeregisterBackends(ids []instance.ID) (loadbalancer.Result, error) {
+	lbLogger.Debug("DeregisterBackends", "name", l.name, "ids", ids, "V", debugV)
+	l.lock.Lock()
+	defer l.lock.Unlock()
+
+	// Output array of members removed
+	membersToRemove := []string{}
+
+	// Get current list of backend members
+	lb, err := l.lbService.Mask("members").GetLoadBalancer(&l.uuid)
+	if err != nil {
+		return nil, err
+	}
+
+	// Iterate over the list of members to remove checking if they exist in the loadbalancer.
+	// If they do, add member to removal list. If not, just log it and continue.
+	for _, id := range ids {
+		ipFound := false
+		ip := string(id)
+		for _, member := range lb.Members {
+			if *member.Address == ip {
+				membersToRemove = append(membersToRemove, *member.Uuid)
+				ipFound = true
+				break
+			}
+		}
+		if ipFound == false {
+			lbLogger.Error("DeregisterBackends", "name", l.name, "V", debugV, "backend not found", ip)
+		}
+	}
+	// If there are any members to remove, remove them.
+	if len(membersToRemove) > 0 {
+		attempt := 0
+		for {
+			attempt++
+			_, err = l.lbMemberService.DeleteLoadBalancerMembers(&l.uuid, membersToRemove)
+			if err == nil {
+				break
+			}
+			// Check for the lb is in state UPDATE_PENDING. (HTTP 500)
+			if strings.Contains(err.Error(), "UPDATE_PENDING") && attempt <= updateRetryCount {
+				time.Sleep(updateRetryTime)
+			} else {
+				return nil, err
+			}
+		}
+	}
+
+	lbLogger.Debug("DeregisterBackends", "name", l.name, "V", debugV, "backends", membersToRemove)
+	return lbResult("deregister"), nil
+}
+
+// Backends returns a list of backends
```

Comment:
Can we explain that we are using the IP address as a common correlation key? Even though this func is returning an `[]instance.ID`, the value is simply the IP address of each member, right?

Reply:
Correct.

Reply:
I'll add a bigger remark that we are using the private IP address of the worker instead of the worker name. This was because there wasn't a field to map worker name to ip within the LBaaS APIs.